File: backend/src/main/java/com/cheremushkin/data/WorkEvent.java
package com.cheremushkin.data;

import com.cheremushkin.serializer.WorkEventSerializer;
import com.esotericsoftware.kryo.DefaultSerializer;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.extern.jackson.Jacksonized;

import java.time.ZonedDateTime;
import java.util.Map;

/**
 * Work event. Both come from UI and internal server events
 * are in this class
 */
@Data
@Builder
@Jacksonized
@AllArgsConstructor
@DefaultSerializer(WorkEventSerializer.class)
final public class WorkEvent {
    /**
     * Fired in: UI
     * Consumed in: ValidateFunction
     *
     * Sent when UI is started.
     * Event: Is sent to backend when UI is started, that means that user enters url in browser.
     * Parameters: SESSION_KEY - session key. Can be empty or
     * contain 8 hex digits.
     * Caller: backend.service
     *
     * Sent to backend when UI starts. Contains session key or empty
     * for new session.
     *
     * When user first time enter the site, key is empty. New key is obtained and UI is redirected.
     *
     * When user login to site not for the first time, key is already there. User saved the page
     * to bookmark with a key. All cards and settings are kept for this key.
     */
    public static final String UI_START_EVENT = "UI_START_EVENT";

    public static final String SESSION_KEY = "SESSION_KEY";

    /**
     * Internal Backend events
     */

    /**
     * Fired in: ValidateFunction
     * Consumed in: MainFunction
     * Parameters: SESSION_KEY - session key, generated in ValidationFunction
     * Sent when UI is started without key.
     * New UI session is started. Session id is generated on backend
     * and sent to UI. New key in backend map will be created.
     * UI will be redirected to new url with session key.
     */
    public static final String UI_START_WITHOUT_KEY_EVENT = "UI_START_WITHOUT_KEY_EVENT";
    /**
     * Fired in: ValidateFunction
     * Consumed in: MainFunction
     * Start UI with existing key. Key is validated. Map records are updated
     * with new login time.
     */
    public static final String UI_START_WITH_KEY_EVENT = "UI_START_WITH_KEY_EVENT";
    /**
     * Fired in: MainFunction
     * Consumed in: UI
     * Fired when new session is created. New session key is stored in backend and fired back to UI.
     * New session is created.
     */
    public static final String BACKEND_NEW_KEY_EVENT = "BACKEND_NEW_KEY_EVENT";
    /**
     * Fired in: UI
     * Consumed by: MainFunction
     * When create card button is pressed then this event arise. This event
     * requests card id from backend. Card id is same as key: 8 hex digits.
     */
    public static final String CARD_GET_ID_EVENT = "CARD_GET_ID_EVENT";
    /**
     * Fired in: MainFunction
     * Consumed by: UI
     * Arise in main function when new id is generated. After that it is send to UI
     */
    public static final String BACKEND_NEW_ID_EVENT = "BACKEND_NEW_ID_EVENT";
    /**
     * New card id. Is sent with BACKEND_NEW_ID_EVENT to UI.
     */
    public static final String ID = "ID";
    /**
     * Fired in: UI
     * Consumed in: MainFunction
     * Fired when card is updated. For example coordinates or
     * content is changed.
     */
    public static final String UPDATE_CARD_EVENT = "UPDATE_CARD_EVENT";
    /**
     * Fired in: UI
     * Consumed in; MainFunction
     * Delete card - move it to done. All cards are not deleted.
     * They are moved to archive.
     * Event data: ID = card id.
     */
    public static final String DELETE_CARD_EVENT = "DELETE_CARD_EVENT";

    /**
     * Card property of data
     */
    public static final String CARD = "CARD";

    /**
     * Error event, should be passed to UI when error occurs.
     * For example session key is invalid
     */
    private static final String ERROR_EVENT = "ERROR_EVENT";

    /**
     * Type of event. Returned from UI. Also filled on backend.
     * Backend specific events were here: Internal Backend events
     */
    String type;
    /**
     * Create time.
     */
    ZonedDateTime dateTime;
    /**
     * Data that is in event
     */
    Map<String, String> data;

    public static WorkEvent buildErrorEvent() {
        return WorkEvent.builder().type(WorkEvent.ERROR_EVENT).build();
    }

    public WorkEvent add(String key, String value) {
        data.put(key, value);
        return this;
    }

    public static WorkEventBuilder builder() {
        return new WorkEventCustomBuilder();
    }

    private static class WorkEventCustomBuilder extends WorkEventBuilder {
        @Override
        public WorkEvent build() {
            if (super.dateTime == null) super.dateTime = ZonedDateTime.now();
            return super.build();
        }
    }
}

File: app/src/app/data/work-event.ts
import { TSMap } from "typescript-map";

/**
 * Represents work event. Work event is event
 * with type and optional data to exchange
 * events inside fronend and on backend.
 */
export class WorkEvent {

  /**
   * New card events
   */

  /**
   * Add new card event. New card is allowed to be added.
   * See diagrams at diagrams/ folder for details.
   * New card button is pressed. After that allow service is 
   * checked. If allow service allow adding new card, this event occurs.
   * Sender: add.component
   * Receiver: card.service
   */
  static ALLOW_ADD_EVENT = "ALLOW_ADD_EVENT"
  
  /**
   * Event: Get new card Id from backend
   * Parameters: none
   * Sender: card.service
   * 
   * Sent when new card event come to card.service. Card service
   * requests new id from backend. Session key will be added in 
   * routing service to this event to send it to backend.
   */
  static CARD_GET_ID_EVENT = "CARD_GET_ID_EVENT"


  /**
   * Navigation events
   * Event: End user navigation
   * Parameters: SESSION_KEY - session key
   * Sender: app.component
   * 
   * Called when user enter URL in browser. URL can contain
   * session key (16 hex digits number). URL is in form 
   * worktask.io/key . URL may not contain key: worktask.io
   * In this case session is new. We obtain new session from 
   * server and redirect to URL worktask.io/new_key
   */
  static APP_NAVIGATION_END_EVENT = "APP_NAVIGATION_END_EVENT"
  static SESSION_KEY = "SESSION_KEY"

  /**
   * Sent when UI is started.
   * Event: UI is sent to backend when UI is started. User
   * enter url in browser.
   * Parameters: SESSION_KEY - session key. Can be empty or
   * contain 16 hex digits.
   * Sender: backend.service
   * 
   * Sent to backend when UI starts. Contains session key or empty
   * for new session. When user first time enter the site. New key
   * is obtained and UI is redirected.
   */
  static UI_START_EVENT = "UI_START_EVENT"

  /**
   * Update card event. Sent when card is saved
   * or when card is finished dragging.
   * Parameters: CARD - json card to save
   * Sender: card.component
   * Receiver: Backend
   * 
   * Send card information to backend
   */
  static UPDATE_CARD_EVENT = "UPDATE_CARD_EVENT"
  static CARD = "CARD"

  /**
   * Sent when drag start to be enabled
   * Sender: allow.service
   * Receiver: card.component
   */
  static DRAG_ENABLED_EVENT = "DRAG_ENABLED_EVENT"

  /**
   * Sent when drag start to be disabled
   * Sender: allow.service
   * Receiver: card.component
   */
  static DRAG_DISABLED_EVENT = "DRAG_DISABLED_EVENT"

  /**
   * Event with new key from backend. Sent when new backend key 
   * is generated.
   * 
   * Frontend use first time enters UI. Then backend receives UI_START_EVENT.
   * Backend generates new key and sent BACKEND_NEW_KEY_EVENT to frontend. 
   * 
   * Sender: Backend
   * Receiver: routing.service
   * Parameters: SESSION_KEY - session key
   */
  static BACKEND_NEW_KEY_EVENT = "BACKEND_NEW_KEY_EVENT"

  /**
   * Generated on backend. Send new card key to frontend.
   * 
   * Sender: backend
   * Receiver: card.service
   * Parameters: ID - new card id
   */
  static BACKEND_NEW_ID_EVENT = "BACKEND_NEW_ID_EVENT"

  /**
   * Generated when card is done. Done means that card disappears from 
   * frontend.
   * Sender: card.component
   * Receiver: backend.service
   * Parameters: ID - card id
   */
  static DONE_CARD_EVENT = "DONE_CARD_EVENT"
  static ID = "ID"

  /**
   * WorkEvent type
   */
  public type: string;
  /**
   * Creation date time
   */
  public createDate: Date;
  /**
   * Data in for of string key-value pairs
   */
  public data: TSMap<string, string>;

  /**
   * Constructs a WorkEvent. WorkEvent contains
   * event id and properties as string map.
   * Each property has name and value.
   * 
   * @param type event type. See static types of 
   * events in this class.
   * @param args each argument is either key or value.
   * First argument is a key, second is value,
   * the again key, value, etc. There should be
   * event number of arguments. See static types of keys
   * in this class
   * @returns new WorkEvent
   */
  constructor(
    type: string,
    ...args: string[] | WorkEvent[]
    ) {
        this.type = type
        this.createDate = new Date()
        this.data = new TSMap<string, any>()
        if (args.length == 0) return
        if (args[0] instanceof WorkEvent) {
          const event = args[0]
          event.data.forEach(
            (value: string, key?: string) => {
              if (key) this.data.set(key, value)
            })
        } else {
          for(var i = 0; i < args.length; i +=2) {
            this.data.set(args[i] as string, args[i+1] as string)
          }
        }
  }
}